staging/src/k8s.io/client-go/examples/controller-ref-manager/main.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"os"
+
+	apps "k8s.io/api/apps/v1beta2"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	// Uncomment the following line to load the gcp plugin (only required to authenticate against GKE clusters).
+	// _ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+
+	"k8s.io/client-go/examples/controller-ref-manager/controller"
+)
+
+var (
+	productionLabel         = map[string]string{"type": "production", "app": "demo"}
+	productionLabelSelector = labels.Set{"type": "production"}.AsSelector()
+)
+
+// RefManager normally used in controllers sync loop to adopt/release resources.
+// See https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/controller-ref.md
+// for more details.
+func main() {
+	kubeconfig := flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
+	flag.Parse()
+	if *kubeconfig == "" {
+		panic("-kubeconfig not specified")
+	}
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		panic(err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err)
+	}
+
+	myController := controller.MyController{ObjectMeta: metav1.ObjectMeta{Name: "demo"}}
+	myController.UID = types.UID("123")
+
+	var d *apps.Deployment
+
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum (see #42639).
+	canAdoptFunc := RecheckDeletionTimestamp(func() (metav1.Object, error) {
+		fresh, err := clientset.AppsV1beta1().Deployments(d.Namespace).Get(d.Name, metav1.GetOptions{})
```

Comment:
odd that this is a closure around `d`. Isn't there a variant where `d` is passed as an argument?

Reply:
Nope. It used in such way almost in every controller. See: https://github.com/kubernetes/kubernetes/blob/ebe84213531894a1a87c082976bded2705388b71/pkg/controller/daemon/daemon_controller.go#L747 for example.